library/Faktory/Job.hs

```diff
@@ -85,11 +89,18 @@ newJob arg = do
     , jobAt = Nothing
     , jobArgs = pure arg
     , jobOptions = jobtype "Default"
+    , jobFailure = Nothing
     }
 
 jobArg :: Job arg -> arg
 jobArg Job {..} = NE.head jobArgs
 
+jobRetriesRemaining :: Job arg -> Int
+jobRetriesRemaining job = max 0 $ enqueuedRetry - occurredRetries
+  where
+    enqueuedRetry = maybe faktoryDefaultRetry getLast $ joRetry $ jobOptions job
+    occurredRetries = maybe 0 ((+1) . jfRetryCount) $ jobFailure job
```

Comment:
is this `occurredRetries` or `occurredTries`? The +1 is weird, but based on the format, I don't know that you can get rid of it.

Reply:
Yeah that +1 is weird, but very important, see https://github.com/contribsys/faktory/issues/374#issuecomment-902075572.

The confusion (to me) is that `jobs.failure.retry_count` is the count of retries _not including this retry itself_. And in order to compute retries-remaining we need to include this retry itself, and only in the non-null case (hence `maybe 0 ((+1) .`).

It's most like `occurredRetriesIncludingThisVeryOne`, I suppose.

```
1 Attempt: retry=2 - (retry_count=  ?occurred_retries?=0) -> remaining=2
2 Retry 1: retry=2 - (retry_count=0 ?occurred_retries?=1) -> remaining=1
3 Retry 2: retry=2 - (retry_count=1 ?occurred_retries?=2) -> remaining=0
```

:shrug:

Reply:
:eye_twitch: